SkeletonUtils: Add clone() helper for skinned meshes.

donmccurdy opened the pull request:
Fixes https://github.com/mrdoob/three.js/issues/5878, fixes https://github.com/mrdoob/three.js/issues/11574.

Adds a helper method, `SkeletonUtils.clone( object )` that preserves SkinnedMesh/Bone relationships. Ideally this would "just work" with the usual `object.clone()`, but I can't see any clean way of doing that — bones are not necessarily children of the SkinnedMesh(es?) they're associated with, and trying to connect them requires logic that would need to live somewhere outside the recursive callback series. This helper method provides an easy way to work around the problem.

Based on https://gist.github.com/zellski/be4e9207ab8e70c4e89062d48ce345ba#file-gltf_utils-js-L19 by @zellski and https://gist.github.com/cdata/f2d7a6ccdec071839bc1954c32595e87 by @cdata.

Commit:
AnimationUtils: Add clone() helper for skinned meshes.

AnimationUtils: Clean up lint errors.

AnimationUtils: Ensure .clone() does not depend on node names.

GLTFLoader: Revert automatic node naming.

Mugen87 commented:
I think it's good to have this function as a work around until we solve the real problem and have a consistent approach for skeleton animation in `three.js`.

takahirox commented:
I haven't follow up the discussion yet, but why is cloning method for `SkinnedMesh` in `AnimationUtils`?

takahirox commented on src/animation/AnimationUtils.js:
```diff
@@ -158,6 +158,50 @@ var AnimationUtils = {
+
+		source.traverse( function ( node ) {
+
+			sourceNodes[ node.name ] = node;
```
Do you assume node has an unique name?

donmccurdy replied:
Unique within the subtree being cloned, yes.

donmccurdy commented:
> I haven't followed up the discussion yet, but why is cloning method for SkinnedMesh in AnimationUtils?

I'd rather it just work via `mesh.clone()` of course, but bones may not be children of the mesh. Consider this subtree:

- object
- object
- object
  - bone
    - bone
      - bone
- object
  - skinned mesh

There's no guarantee the new bones will exist at the time the skinned mesh is recursively cloned. It seems like we need somewhere else to put that logic, and so a Utils class seems as good as any. [This solution](https://gist.github.com/zellski/be4e9207ab8e70c4e89062d48ce345ba#file-gltf_utils-js-L19) is another, but relies on cloning happening at the Scene level, and not all loaders return Scene objects.

takahirox commented:
Sorry I wrote poorly. I think it's good to have a workaround in Utils for now. My question is why is it in `"Animation"Utils` tho cloning `SkinnedMesh` isn't animation special.

takahirox commented on src/animation/AnimationUtils.js:
```diff
@@ -158,6 +158,50 @@ var AnimationUtils = {
+
+			clonedMesh.skeleton.bones = sourceBones.map( function ( bone ) {
+
+				return clonedNodes[ bone.name ];
```
Similarly do you assume bones are in a subtree?

donmccurdy replied:
This assumes the bones are within the subtree of the object being cloned, but not necessarily descendants of the SkinnedMesh they're associated with.

takahirox replied:
I think we need to check if bones are in a subtree and halt with warn/error log if they aren't. Because it isn't guaranteed, for example users might call `Utils.clone( object2 )` for the following graph. Sometimes even bones wouldn't be in the scene graph.

- object1
  - bone1
    - bone2
- object2
  - skinned mesh

donmccurdy commented:
> My question is why is it in "Animation"Utils tho cloning SkinnedMesh isn't animation special.

Because I can't think of a better place to put it, and skinning is at least somewhat related to animation. Suggestions welcome. 😉

takahirox commented:
Hm, `SceneUtils` or `(Skinned)MeshUtils`?

donmccurdy commented on examples/js/loaders/GLTFLoader.js:
```diff
@@ -2979,6 +2981,10 @@ THREE.GLTFLoader = ( function () {
 
+			} else {
+
+				node.name = 'gltfNode' + ( nextNodeIndex++ );
```
Right. 👍

takahirox commented on src/animation/AnimationUtils.js:
```diff
@@ -158,6 +158,50 @@ var AnimationUtils = {
+			var sourceBones = sourceMesh.skeleton.bones;
+
+			clonedMesh.skeleton = sourceMesh.skeleton.clone();
+			clonedMesh.bindMatrix = sourceMesh.bindMatrix;
```
We should copy the matrix, rather than we share?

donmccurdy replied:
Done ✅

donmccurdy commented:
Updated to ensure cloning does not depend on names. I removed the part adding unique names to glTF nodes for now, but will probably try another PR for that in the future since animation tracks still require it.

takahirox commented on src/animation/AnimationUtils.js:
```diff
@@ -158,9 +158,59 @@ var AnimationUtils = {
+
+			} );
+
+			clonedMesh.bind( clonedMesh.skeleton, clonedMesh.bindMatrix );
```
Copying matrix to `.bindMatrix` is done in `SkinnedMesh.bind()`.

https://github.com/mrdoob/three.js/blob/dev/src/objects/SkinnedMesh.js#L112

So I think you can replace this line with

    clonedMesh.bind( clonedMesh.skeleton, sourceMesh.bindMatrix );

and remove the line above `clonedMesh.bindMatrix.copy( sourceMesh.bindMatrix );`.

donmccurdy replied:
Done ✅

takahirox replied:
You need to pass `"source"Mesh.bindMatrix`.

donmccurdy replied:
Fixed ✅

takahirox commented on src/animation/AnimationUtils.js:
```diff
@@ -158,9 +158,58 @@ var AnimationUtils = {
+		} );
+
+		return clone;
+
```
Can't we remove `sourceLookup`?

    source.traverse( function ( node ) {

        if ( ! node.isSkinnedMesh ) return;

        var sourceMesh = node;
        var sourceBones = sourceMesh.skeleton.bones;
        var clonedMesh = cloneLookup.get( node );

        clonedMesh.skeleton = sourceMesh.skeleton.clone();

        clonedMesh.skeleton.bones = sourceBones.map( function ( bone ) {

            return cloneLookup.get( bone );

        } );

        clonedMesh.bind( clonedMesh.skeleton, sourceMesh.bindMatrix );

    } );

    return clone;

donmccurdy replied:
Fixed ✅

takahirox commented:
Thanks for the update.

So, you still want this method in `AnimationUtils`? Or do you move it to `SceneUtils` or somewhere?

takahirox commented on src/animation/AnimationUtils.js:
```diff
@@ -158,9 +158,55 @@ var AnimationUtils = {
+
+			clonedMesh.skeleton.bones = sourceBones.map( function ( bone ) {
+
+				return cloneLookup.get( bone );
```
Do you wanna check the existence of bones in the subtree as I mentioned?

donmccurdy replied:
Done ✅

donmccurdy commented:
I'm OK with whatever namespace for this, but I don't think I have a better idea than the current name... maybe `cloneWithBones` or `cloneWithSkinning`? It is all a less-preferable solution than fixing `.clone()` directly but I don't see any straightforward way to do that. /cc @mrdoob

takahirox commented on src/animation/AnimationUtils.js:
```diff
@@ -158,9 +158,63 @@ var AnimationUtils = {
+
+					throw new Error( 'THREE.AnimationUtils: Required bones are not descendants of the given object.' );
+
+				}
```
nit: Using `.has()` may look more natural?

    if ( ! cloneLookup.has( sourceBone ) ) {

        throw new Error( 'THREE.AnimationUtils: Required bones are not descendants of the given object.' );

    }

    var clonedBone = cloneLookup.get( sourceBone );

donmccurdy replied:
Done ✅

takahirox commented:
Voting for `SceneUtils.cloneWithSkinning`.

donmccurdy commented:
If https://github.com/mrdoob/three.js/pull/14574 goes in, SkeletonUtils seems like a good place to put this. 🤔

takahirox commented:
Agreed.

mrdoob commented:
> If https://github.com/mrdoob/three.js/pull/14574 goes in, SkeletonUtils seems like a good place to put this. 🤔

Merged.

donmccurdy commented:
@takahirox @mrdoob Does `SkeletonUtils.clone()` seem OK as a name? Or something like more explicit like `.cloneWithSkinning()`?

takahirox commented:
I prefer the explicit one.

Mugen87 commented:
I have lately used this code in a project and it works great. Would love to see it in the code base :blush:

Eketol commented:
Me too. I've also tested it and it works fine for me. Just one question: the use of Map() is "es2015" (ES6), is that ok?
Also, in case it helps other noobs like me: Materials are not cloned, but referenced. In my case I think it is the right behaviour. If you want to customise some of the cloned object's materials, bear that in mind.

Mugen87 commented:
I think using `Map` in this case is okay since it's no part of the core.

donmccurdy committed:
Move 'clone' from AnimationUtils to SkeletonUtils.

donmccurdy commented:
Thanks!

I've moved the `clone()` method from AnimationUtils to SkeletonUtils, and added documentation for SkeletonUtils. I think this is ready to go. 👍

donmccurdy commented:
@mrdoob could this be included with r103?

titansoftime commented:
This just solved a huge issue I was having. Thank you.

Mugen87 commented:
> This just solved a huge issue I was having.

https://discourse.threejs.org/t/solved-issue-with-gltfloader-and-reusing-geometry/6697

mrdoob commented:
> @mrdoob could this be included with r103?

Done! 😇

mrdoob commented:
Thanks!

canadaduane commented:
> Also, in case it helps other noobs like me: Materials are not cloned, but referenced. In my case I think it is the right behaviour. If you want to customise some of the cloned object's materials, bear that in mind.

Is there an option to clone the materials as well? In my case, I'd like some of my skinnedmesh objects to be transparent (but not all), even though they've been cloned from the same original.

Mugen87 commented:
`SkeletonUtils.clone()` does not touch materials. You have to apply the material changes in a subsequent step.

enricocantori commented:
Three.is is frequently used in the world of character animations. I can’t find no functions, no working examples related to the replace of parts of characters. Almost all characters are composed by skinnedMesh children and a lot of time is necessary change a part of the character like clothes, hair, face within characters that have same skeleton without broke animation (ex. Readyplayes characters). Clone and SkeletonUtils.Clone doesn’t solve this problem. Exist one more specific function?
